src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2061,44 +2050,9 @@ RunDistributedExecution(DistributedExecution *execution)
 			/* wait for I/O events */
 			int eventCount = WaitEventSetWait(execution->waitEventSet, timeout, events,
 											  eventSetSize, WAIT_EVENT_CLIENT_READ);
-
-			/* process I/O events */
-			for (; eventIndex < eventCount; eventIndex++)
+			if (!ProcessWaitEvents(execution, events, eventCount))
 			{
-				WaitEvent *event = &events[eventIndex];
-
-				if (event->events & WL_POSTMASTER_DEATH)
-				{
-					ereport(ERROR, (errmsg("postmaster was shut down, exiting")));
-				}
-
-				if (event->events & WL_LATCH_SET)
-				{
-					ResetLatch(MyLatch);
-
-					if (execution->raiseInterrupts)
-					{
-						CHECK_FOR_INTERRUPTS();
-					}
-
-					if (IsHoldOffCancellationReceived())
-					{
-						/*
-						 * Break out of event loop immediately in case of cancellation.
-						 * We cannot use "return" here inside a PG_TRY() block since
-						 * then the exception stack won't be reset.
-						 */
-						cancellationReceived = true;
-						break;
-					}
-
-					continue;
-				}
-
-				WorkerSession *session = (WorkerSession *) event->user_data;
-				session->latestUnconsumedWaitEvents = event->events;
-
-				ConnectionStateMachine(session);
+				cancellationReceived = true;
```

Comment:
It would be a little nicer to pass this by pointer. It's not entirely obvious that the return value of ProcessWaitEvents indicates cancellation.

```c
ProcessWaitEvents(execution, events, eventCount, &cancellationReceived);
```